rfcs/QueryLevelNullability.md

```diff
@@ -170,5 +170,19 @@ This syntax consciously does not cover the following use cases:
   that says "if a value would be `null` make it some other value instead". We are
   not interested in covering that in this proposal.
   
-## Implementation
-GraphQL.js: https://github.com/graphql/graphql-js/pull/2824
+## Work Items
+Patches that will need to be made if this proposal is accepted. The 
+[RFC proposal process](https://github.com/graphql/graphql-spec/blob/main/CONTRIBUTING.md)
+requires that a proof of concept is implemented in a GraphQL library. As work items are completed,
+PRs will be linked here.
+- Spec Changes
+- Official Libraries
+  - GraphQL.js: https://github.com/graphql/graphql-js/pull/2824
+- 3rd Party Libraries
+  - [Apollo Android](https://github.com/apollographql/apollo-android)
```

Comment:
We need the same for Apollo JS Client at https://github.com/apollographql/apollo-client

Reply:
I expect these tests to pass. With @apollo/client at 3.3.11, the first test passes but the second test fails as the `readQuery` returns `null`:

```tsx
import { gql, InMemoryCache } from '@apollo/client';

const nonNullQuery = gql`
  query getToy($id: ID!) {
    toy(id: $id) {
      id
      name @nonNull
    }
  }
`;

const nullableQuery = gql`
  query getToy($id: ID!) {
    toy(id: $id) {
      id
      name
    }
  }
`;

describe('@nonNull directive', () => {
  it('returns null for the entire query when the cached value is null and the field is marked @nonNull', () => {
    const cache = new InMemoryCache();

    cache.writeQuery({
      query: nullableQuery,
      variables: { id: '12345' },
      data: {
        toy: {
          __typename: 'Toy',
          id: '12345',
          name: null,
        },
      },
    });

    expect(
      cache.readQuery({
        query: nonNullQuery,
        variables: { id: '12345' },
      })
    ).toBeNull();
  });

  it('returns the cached value for the entire query when the cached value is non-null and the field is marked @nonNull', () => {
    const cache = new InMemoryCache();

    cache.writeQuery({
      query: nullableQuery,
      variables: { id: '12345' },
      data: {
        toy: {
          __typename: 'Toy',
          id: '12345',
          name: 'Name',
        },
      },
    });

    expect(
      cache.readQuery({
        query: nonNullQuery,
        variables: { id: '12345' },
      })
    ).toEqual({
      toy: {
        __typename: 'Toy',
        id: '12345',
        name: 'Name',
      },
    });
  });
});
```

Reply:
Thanks for writing up the tests! We'll have to come back and get this when we start implementation.